File: CodingBlocks/AlgoSummerJune19/Session4/Quicksort.cpp
#include <iostream>
using namespace std;

int partition(int *arr, int start, int end) {
	int i = start-1;
	int j = start;
	int pivot = arr[end];
	while (j < end) {
		if (arr[j] <= pivot) {
			i++;
			swap(arr[i], arr[j]);
		}
		j++;
	}
	swap(arr[i+1], arr[end]);
	return i+1;
}

void quickSortHelper(int *arr, int start, int end) {
	if(start >= end) {
		return;
	}

	int pivot = partition(arr, start, end);
	quickSortHelper(arr, start, pivot - 1);
	quickSortHelper(arr, pivot+1, end);
}

void quickSort(int *arr, int n) {
	quickSortHelper(arr, 0, n - 1);
}

int main(int argc, char const *argv[])
{
	int n;
	cin>>n;
	int *arr = new int[n];
	for(int i=0;i<n;i++) {
		cin>>arr[i];
	}
	quickSort(arr, n);
	for(int i=0;i<n;i++) {
		cout<<arr[i]<<" ";
	}

	return 0;
}

File: UnacademyCourse/SortingAndSearchingPythonJune2021/Lecture3/Quick_Sort.py
from random import randint
def partition(arr, lo, hi) -> int:
    p = randint(lo, hi)
    arr[hi], arr[p] = arr[p], arr[hi]
    pivot = arr[hi]
    m = lo-1
    for i in range(lo, hi):
        if arr[i] <= pivot:
            m = m+1
            arr[i], arr[m] = arr[m], arr[i]
        
    arr[m+1], arr[hi] = arr[hi], arr[m+1]
    return m+1


def quick_sort_helper(arr, lo, hi):
    if lo < hi:
        pi = partition(arr, lo, hi)
        quick_sort_helper(arr, lo, pi-1)
        quick_sort_helper(arr,pi+1, hi)


def quick_sort(arr):
    quick_sort_helper(arr, 0, len(arr) - 1)




arr = [10,7,8,9,1,5]
quick_sort(arr)
print(arr)